File: lib/AdventOfCode/Solution/Day06.ex
defmodule Solutions.Day06 do
  alias Helpers.Readers, as: Reader

  def genChunker(len) do
    chunk_fun = fn element, acc ->
      if length(acc) == len do
        {:cont, Enum.reverse(acc), [element | Enum.drop(acc, -1)]}
      else
        {:cont, [element | acc]}
      end
    end
    after_fun = fn
      [] -> {:cont, []}
      acc -> {:cont, Enum.reverse(acc), []}
    end
    [chunk_fun, after_fun]
  end

  def getPacketStart(data, len) do
    [chunker, afterfun] = genChunker(len)
    Enum.map(data, &String.to_charlist/1) |> Enum.map(fn x-> Enum.chunk_while(x, [], chunker, afterfun) end) |>
      Enum.map(fn row ->
        Enum.map(row, fn entry ->
          MapSet.new(entry) |> Enum.count
        end)
      end)
    |> Enum.map(fn row ->
      Enum.find_index(row, fn x -> x == len end) + len
    end) |> List.first
  end

  def run(args) do
    data = Reader.readFile(args)
    getPacketStart(data, 4)
    |> IO.inspect(label: "Task 01")
    getPacketStart(data, 14)
    |> IO.inspect(label: "Task 02")
  end
end

File: src/main/scala/Solutions/Day06.scala
package AoC2022
package Solutions

import Helpers.Readers

object Day06 {
  def main(args: Array[String]): Unit = {
    val data = Readers.readFile("Data/Day06.txt").head
    val calc = (len: Int) => LazyList.from(data).sliding(len).takeWhile(it => Set.from(it).size < len).size + len
    println("Task 01: " + calc(4))
    println("Task 02: " + calc(14))
  }
}
